packages/react/focus-scope/src/FocusScope.tsx

```diff
@@ -26,44 +35,265 @@ type FocusScopeProps = {
+  return [first, last] as const;
+}
+
+/**
+ * Returns a list of potential tabbable candidates.
+ * NOTE: They are considered candidates at this point, because it includes elements that are hidden.
+ */
+function getTabbableCandidates(container: HTMLElement) {
+  const nodes: HTMLElement[] = [];
+  const walker = createTabbableWalker(container);
+  while (walker.nextNode()) nodes.push(walker.currentNode as HTMLElement);
+  // we do not take into account the order of nodes with positive `tabIndex` as it is
+  // considered an accessibility issue.
+  return nodes;
+}
+
+/**
+ * Creates a tree that contains only tabbable elements.
+ * This is only a close aproximation. For example it doesn't take into account cases like when
+ * elements are not visible. This cannot be worked out easily by just reading a property, but rather
+ * necessitate runtime knowledge (computed styles, etc). We deal with these cases separately.
+ *
+ * See: https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker
+ * Credit: https://github.com/discord/focus-layers/blob/master/src/util/wrapFocus.tsx#L1
+ */
+function createTabbableWalker(container: HTMLElement) {
+  return document.createTreeWalker(container, NodeFilter.SHOW_ELEMENT, {
+    acceptNode: (node: any) => {
+      const isHiddenInput = node.tagName === 'INPUT' && node.type === 'hidden';
+      if (node.disabled || node.hidden || isHiddenInput) return NodeFilter.FILTER_SKIP;
+      // `.tabIndex` is not the same as the `tabindex` attribute. It works on the
+      // runtime's understanding of tabbability, so this automatically accounts
+      // for any kind of element that could be tabbed to.
+      return node.tabIndex >= 0 ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;
+    },
+  });
+}
+
+/**
+ * Returns the first "real" tabbable in a list of candidates.
+ * NOTE: This takes into account if elements are hidden or not.
+ */
+function getFirstRealTabbable(candidates: HTMLElement[], container: HTMLElement) {
+  for (const candidate of candidates) {
+    // we stop checking if it's hidden at the `container` level (excluding)
+    if (!isHidden(candidate, { upTo: container })) return candidate;
+  }
+}
+
+function isHidden(node: HTMLElement, { upTo }: { upTo?: HTMLElement }) {
+  if (getComputedStyle(node).visibility === 'hidden') return true;
+  while (node) {
+    // we stop at `upTo` (excluding it)
+    if (upTo !== undefined && node === upTo) return false;
+    if (getComputedStyle(node).display === 'none') return true;
+    node = node.parentElement as HTMLElement;
+  }
+  return false;
+}
+
+function isSelectableInput(element: any): element is FocusableTarget & { select: () => void } {
+  return element instanceof HTMLInputElement && 'select' in element;
+}
+
+function focus(element?: FocusableTarget | null, { select = false } = {}) {
+  // only focus if that element is focusable and not already focused
+  if (element && element.focus && element !== document.activeElement) {
```

Comment:
curious why we need the `element !== document.activeElement` check here. AFAIK, browsers won't fire focus for something that is already focused. Is there another reason here that we should maybe document?

Reply:
Hmm, yeah that's suspicious, although I clearly remember showing you that on a call and that it was indeed a small improvement on what we had before.

It might be for one of these edge cases where the whole page gets blurred and focus gets returned to the same place.
I will double-check that.

Edit: on second thoughts, I might also have added this to avoid "selection" as the browser would run this even if it was focused already. Again, I'll double check.